src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h

```diff
@@ -0,0 +1,1777 @@
+      local_move_back(node_key, iter + 1, iter, iter_end());
+
+    iter->set_node_key(node_key);
+    set_size(get_size() + 1);
+    iter->set_node_val(key);
+    iter->set_string_val(val);
+  }
+
+  void leaf_update(
+    iterator iter,
+    const std::string &key,
+    const std::string &val) {
+    assert(iter != iter_end());
+    if (VALIDATE_INVARIANTS) {
+      assert(is_overflow(0, val.size() + 1) == false);
+    }
+    leaf_remove(iter);
+    leaf_insert(iter, key, val);
+  }
+
+  void leaf_remove(iterator iter) {
+    assert(iter != iter_end());
+    if ((iter + 1) != iter_end())
+      local_move_ahead(iter, iter + 1, iter_end());
+    set_size(get_size() - 1);
+  }
+
+  /**
+   * get_key_ptr
+   *
+   * Get pointer to start of key array
+   */
+  omap_leaf_key_le_t *get_node_key_ptr() {
+    return L::Partial(1, 1, get_size()).template Pointer<2>(buf);
+  }
+  const omap_leaf_key_le_t *get_node_key_ptr() const {
+    return L::Partial(1, 1, get_size()).template Pointer<2>(buf);
+  }
+
+  /**
+   * copy_from_foreign_head
+   *
+   * Copy from another node begin entries to this node.
+   * [from_src, to_src) is another node entry range.
+   * tgt is this node entry to copy to.
+   * tgt and from_src must be from different nodes.
+   * from_src and to_src must be in the same node.
+   */
+  static void copy_from_foreign_head(
+    iterator tgt,
+    const_iterator from_src,
+    const_iterator to_src) {
+    assert(tgt->node != from_src->node);
+    assert(to_src->node == from_src->node);
+    void* des = tgt.node->from_end((to_src -1)->get_node_key().key_off);
+    void* src = (to_src - 1)->get_node_val_ptr();
+    size_t len = (to_src -1)->get_node_key().key_off;
+    memcpy(des, src, len);
+    memcpy(
+      tgt->get_node_key_ptr(), from_src->get_node_key_ptr(),
+      to_src->get_node_key_ptr() - from_src->get_node_key_ptr());
+  }
+
+  /**
+   * copy_from_foreign_back
+   *
+   * Copy from another node back entries to this node.
+   * [from_src, to_src) is another node entry range.
+   * tgt is this node entry to copy to.
+   * tgt and from_src must be from different nodes.
+   * from_src and to_src must be in the same node.
+   */
+  void copy_from_foreign_back(
+    iterator tgt,
+    const_iterator from_src,
+    const_iterator to_src) {
+    assert(tgt->node != from_src->node);
+    assert(to_src->node == from_src->node);
+    auto offset = from_src.get_index() == 0? 0: (from_src-1)->get_node_key().key_off;
+
+    void* des = tgt.node->from_end((to_src -1)->get_node_key().key_off - offset);
+    void* src = (to_src - 1)->get_node_val_ptr();
+    size_t len = from_src.get_index() == 0? (to_src -1)->get_node_key().key_off:
+                 (from_src-1)->get_node_val_ptr() - (to_src -1)->get_node_val_ptr();
+    memcpy(des, src, len);
+    memcpy(
+      tgt->get_node_key_ptr(), from_src->get_node_key_ptr(),
+      to_src->get_node_key_ptr() - from_src->get_node_key_ptr());
+    if ( from_src.get_index() == 0)
+      return;
```

Comment:
If I'm reading this right, the only difference between copy_from_foreign_back and this is that copy_from_foreign_back assumes that this branch is taken.  If that's the case, I think you should eliminate copy_from_foreign_back and rename this to copy_from_foreign.

Reply:
Actually, take a look at the last commit on https://github.com/athanatos/ceph/tree/sjust/pull/37925/wip-omap -- I reworked the iterator methods a bit and was able to unify all 6 *copy_from* methods into a single templated function.  That should make it considerably simpler.  I think you should be able to use the same iterator machinery to simply the local variants as well.

Also, the patch before it does s/BlockSize/BLOCK_SIZE.